### PR TITLE
Produce request encoding performance improvement

### PR DESCRIPTION
**Pull request by zmstone**
produce request encoding is the most critical code path.
judging by the naive bechmark (light version added as a eunit test),
the time cost to complete the eunit case is improved by about 35%

### CODE REVIEW

File: changelog.md [hunk 1]
```diff
@@ -71,3 +71,5 @@
   - Made send_error Reason more informative
 * 2.3.3
   - consumer group message metadata V3
+* 2.4.0
+  - Improvec produce request encoding
```
Comment (k32) [inline changelog.md]: typo: Improve

**Comment by zmstone** [comment]
@k32 possible to merge this one ?

**Comment by zmstone** [comment]
sorry. hold on.

**Comment by zmstone** [comment]
now @k32 please take a look again.

**Review by k32 — APPROVED** [review]

**Comment by k32** [comment]
Thank you!

**Pull request merged**